### DISCUSSION TITLE
BUG: Coordinate shift after reprojecting from epsg:4326 to epsg:28354

### DISCUSSION THREAD

**Discussion by cratcliff** [geopandas/geopandas#2723] [Q&A]
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of geopandas.

- [ ] (optional) I have confirmed this bug exists on the main branch of geopandas.

---
#### Code Sample
```python

import pandas as pd
import geopandas as gpd

df = pd.read_csv('Sample_Dataset.csv')
gdf = gpd.GeoDataFrame(df, geometry=gpd.points_from_xy(df['QGIS_Lon'], df['QGIS_Lat']),crs=4326)

# reproject by epsg
gdf = gdf.to_crs(epsg=28354)
gdf['GPD_East94']= gdf.geometry.apply(lambda p: p.x)
gdf['GPD_North94'] = gdf.geometry.apply(lambda p: p.y)

# Compare differences
gdf['QGIS-GPD_East'] = gdf['QGIS_East']  - gdf['GPD_East94']
gdf['QGIS-GPD_North'] = gdf['QGIS_North']  - gdf['GPD_North94']

```

#### Problem description
I have detected a shift ( ~0.6 easting, ~1.5 northing) when projecting data from WGS 84 (EPSG:4326) to GDA 1994 MGA zone 54 (EPSG:28354). This shift is not seen when the reprojection is completed using QGIS or ArcMap reprojection tools or when using geopandas within QGIS <details> <summary markdown="span">QGIS  3.18    Python 3.7.0   Geopandas 0.7.0</summary>

[SYSTEM INFO
-----------
python     : 3.7.0 (v3.7.0:1bf9cc5093, Jun 27 2018, 04:59:51) [MSC v.1914 64 bit (AMD64)]
executable : C:\Program Files\QGIS 3.18\bin\qgis-bin.exe
machine    : Windows-10-10.0.19041

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : None
GEOS lib   : None
GDAL       : 3.1.4
GDAL data dir: None
PROJ       : 7.0.1
PROJ data dir: C:\PROGRA~1\QGIS3~1.18\apps\Python37\lib\site-packages\pyproj\proj_dir\share\proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.7.0
pandas     : 1.0.1
fiona      : 1.8.18
numpy      : 1.19.1
shapely    : 1.7.0
rtree      : 0.9.7
pyproj     : 2.6.1.post1
matplotlib : 3.1.3
mapclassify: None
pysal      : None
geopy      : None
psycopg2   : 2.8.4 (dt dec pq3 ext lo64)
</details> 
Attached is a csv file containing sample data with the expected coordinates created by the QGIS and ArcMap projection tools.
The output eastings and northings should match.

[Sample_Dataset.csv](https://github.com/geopandas/geopandas/files/8477229/Sample_Dataset.csv)

#### Output of ``geopandas.show_versions()`` tested
<details> 
<summary markdown="span">Python 3.9.12 Geopandas 0.10.2</summary>

[SYSTEM INFO
-----------
python     : 3.9.12 (main, Apr  4 2022, 05:22:27) [MSC v.1916 64 bit (AMD64)]
executable : C:\Users\xxxx\Miniconda3\envs\pa39_dev_up\python.exe
machine    : Windows-10-10.0.19042-SP0

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : None
GEOS lib   : None
GDAL       : 3.4.2
GDAL data dir: None
PROJ       : 9.0.0
PROJ data dir: C:\Users\xxxx\Miniconda3\envs\pa39_dev_up\Library\share\proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.10.2
pandas     : 1.4.2
fiona      : 1.8.21
numpy      : 1.22.3
shapely    : 1.8.0
rtree      : 1.0.0
pyproj     : 3.3.0
matplotlib : 3.5.1
mapclassify: 2.4.3
geopy      : None
psycopg2   : None
geoalchemy2: None
pyarrow    : None
pygeos     : None
</details>
<details> 
<summary markdown="span">Python 3.7.12 Geopandas 0.7.0</summary>

[SYSTEM INFO
-----------
python     : 3.7.12 | packaged by conda-forge | (default, Oct 26 2021, 05:37:49) [MSC v.1916 64 bit (AMD64)]
executable : C:\Users\XXXX\Miniconda3\envs\pa37_dev\python.exe
machine    : Windows-10-10.0.19041-SP0

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : None
GEOS lib   : None
GDAL       : 3.1.4
GDAL data dir: None
PROJ       : 8.0.1
PROJ data dir: C:\Users\XXXX\Miniconda3\envs\pa37_dev\Library\share\proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.7.0
pandas     : 1.3.4
fiona      : 1.8.18
numpy      : 1.21.4
shapely    : 1.8.0
rtree      : 0.9.7
pyproj     : 3.2.1
matplotlib : 3.5.0
mapclassify: None
pysal      : None
geopy      : None
psycopg2   : None]

</details>

**Comment by martinfleis** [comment]
@snowman2 any idea?

**Comment by snowman2** [comment]
There are three different transformations available:
```python
>>> from pyproj.transformer import TransformerGroup
>>> tg = TransformerGroup(4326, 28354, always_xy=True)
>>> tg.transformers
[<Concatenated Operation Transformer: pipeline>
Description: axis order change (2D) + Inverse of GDA94 to WGS 84 (2) + Map Grid of Australia zone 54
Area of Use:
- name: Australia including Lord Howe Island, Macquarie Island, Ashmore and Cartier Islands, Christmas Island, Cocos (Keeling) Islands, Norfolk Island. All onshore and offshore.
- bounds: (93.41, -60.55, 173.34, -8.47), <Concatenated Operation Transformer: pipeline>
Description: axis order change (2D) + Inverse of GDA94 to WGS 84 (1) + Map Grid of Australia zone 54
Area of Use:
- name: Australia including Lord Howe Island, Macquarie Island, Ashmore and Cartier Islands, Christmas Island, Cocos (Keeling) Islands, Norfolk Island. All onshore and offshore.
- bounds: (93.41, -60.55, 173.34, -8.47), <Concatenated Operation Transformer: pipeline>
Description: axis order change (2D) + Inverse of GDA94 to WGS 84 (3) + Map Grid of Australia zone 54
Area of Use:
- name: Australia - Australian Capital Territory; New South Wales; Northern Territory; Queensland; South Australia; Tasmania; Western Australia; Victoria.
- bounds: (112.85, -43.7, 153.69, -9.86)]
```

PROJ attempts to automatically select what it thinks is the best transformation when using `Transformer` (what geopandas uses by default). However, you can manually choose the one you want with `TransformerGroup`:
```python
>>> xx0, yy0 = tg.transformers[0].transform(df.QGIS_Lon, df.QGIS_Lat)
>>> (df.QGIS_East - xx0).median()
0.6244598758639768
>>> (df.QGIS_North - yy0).median()
1.5168016464449465
>>> xx1, yy1 = tg.transformers[1].transform(df.QGIS_Lon, df.QGIS_Lat)
>>> (df.QGIS_North - xx1).median()
5854361.610242816
>>> (df.QGIS_East - xx1).median()
-0.001678180182352662
>>> (df.QGIS_North - yy1).median()
-0.00016847625374794006
>>> xx2, yy2 = tg.transformers[2].transform(df.QGIS_Lon, df.QGIS_Lat)
>>> (df.QGIS_East - xx2).median()
0.6387803747784346
>>> (df.QGIS_North - yy2).median()
1.5095374234952033
```

Based on this, the transformer at index 1 seems to match what you want:
```python
df['GPD_East94'] = xx1
df['GPD_North94'] = yy1
```

**Comment by cratcliff** [comment]
How is the best transformation selected? 

I'm dealing with precision agriculture data so the shift that I have observed it too big. Is there a way of automatically selecting the most accurate transformation without user interaction?

**Comment by snowman2** [comment]
> How is the best transformation selected?

https://proj.org/operations/operations_computation.html

> Is there a way of automatically selecting the most accurate transformation without user interaction?

There are several ways to attempt to do this. Based on the example above, you could export the desired transformer to a PROJ string and use `Transformer.from_pipeline` to create the transformer.
